Modeleur 1.1: API [ANT-1876]

payetvin opened the pull request.

payetvin committed:
Add solver/optim/api, mip files

payetvin committed:
Add mipObjective, license header mpl

payetvin committed:
Add ptim/ortoolsImpl/mipVariable

payetvin committed:
Correct return for set functions

payetvin committed:
CMakefiles

payetvin committed:
configure ok

payetvin committed:
compile ok

payetvin committed:
format

payetvin committed:
ortools mip solution start

payetvin committed:
mip solution constructor

payetvin committed:
format

payetvin committed:
override, explicit

payetvin committed:
get optimalvalue

payetvin committed:
add a objective field

payetvin committed:
mip objective file, getter for mipvariable

payetvin committed:
private members, virtual public

payetvin committed:
mip constraint

payetvin committed:
format

JasonMarechal25 commented:
Could we have a minimal working example? Either as unit test or a stand alone executable.

payetvin committed:
comment header

payetvin committed:
throw bad cast

pet-mit commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/mipSolution.h:
```diff
@@ -0,0 +1,51 @@
+class OrtoolsMipSolution: virtual public Api::MipSolution
+{
+public:
+    OrtoolsMipSolution(operations_research::MPSolver::ResultStatus responseStatus,
```
I think this should be private too

pet-mit commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/mipSolution.h:
```diff
@@ -0,0 +1,51 @@
+    double getOptimalValue(const Api::MipVariable& var) const override;
+    std::vector<double> getOptimalValue(const std::vector<Api::MipVariable>& vars) const override;
+
+    operations_research::MPObjective* objective_;
```
private

pet-mit commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/mipVariable.h:
```diff
@@ -0,0 +1,53 @@
+    double getLb() override;
+    double getUb() override;
+
+    operations_research::MPVariable* get();
```
this can be private by declaring friend classes

pet-mit replied:
also I think we should rename it getMpVariable

pet-mit commented on src/solver/optim/ortoolsImpl/mipConstraint.cpp:
```diff
@@ -0,0 +1,80 @@
+    auto* mpvar = dynamic_cast<OrtoolsMipVariable*>(var);
+    if (!mpvar)
+    {
+        throw std::bad_cast();
```
add explicit error message

pet-mit commented on src/solver/optim/ortoolsImpl/mipConstraint.cpp:
```diff
@@ -0,0 +1,80 @@
+    auto* mpvar = dynamic_cast<OrtoolsMipVariable*>(var);
+    if (!mpvar)
+    {
+        throw std::bad_cast();
```
same

pet-mit commented on src/solver/optim/ortoolsImpl/mipObjective.cpp:
```diff
@@ -0,0 +1,75 @@
+    auto* mpvar = dynamic_cast<OrtoolsMipVariable*>(var);
+    if (!mpvar)
+    {
+        throw std::bad_cast();
```
same

pet-mit commented on src/solver/optim/ortoolsImpl/mipObjective.cpp:
```diff
@@ -0,0 +1,75 @@
+    auto* mpvar = dynamic_cast<OrtoolsMipVariable*>(var);
+    if (!mpvar)
+    {
+        throw std::bad_cast();
```
same

payetvin committed:
linear solver header start

payetvin committed:
comments

payetvin committed:
linearProblem base

payetvin committed:
linearProblem impl

payetvin committed:
use maps

payetvin committed:
fix compile

payetvin committed:
add to map

payetvin committed:
getvariable

payetvin committed:
constraint, error handling

payetvin committed:
objective in linear solver, solution

payetvin committed:
OrtoolsLinearSolver::solve

payetvin committed:
format

payetvin committed:
remove ;

payetvin committed:
unique ptr for mpSolver_

payetvin committed:
rename to linear problem

payetvin committed:
api linearProblemData

payetvin committed:
base class ortools linearproblemdata

payetvin committed:
fix compile

payetvin committed:
members, methods

payetvin committed:
timer data scalar data

flomnes commented on src/solver/optim/api/include/antares/solver/optim/api/mipVariable.h:
```diff
@@ -0,0 +1,39 @@
+
+class MipVariable
+{
+public:
```
Name ?

flomnes commented on src/solver/optim/api/include/antares/solver/optim/api/mipConstraint.h:
```diff
@@ -0,0 +1,44 @@
+class MipConstraint
+{
+public:
+    virtual void setLb(double lb) = 0;
```
Name ?

flomnes reviewed (commented):
Virtual class for bounds ?
- SetUB
- SetLB
- SetBounds
- GetLB
- GetUB

This can be shared for variables & constraints
```cpp
class HasBounds
{
public
   virtual void SetBounds(double lb, double ub) = 0;
   // etc.
};

class MipVariable : public HasBounds
{
};
class MipConstraint : public HasBounds
{
// ...
};
```

### EDIT
Same for name
```cpp
class HasName {
    virtual const std::string& name() const = 0;
};
```

payetvin committed:
comment unused arg

payetvin committed:
implement linearproblemdata in api

flomnes commented on src/solver/optim/api/include/antares/solver/optim/api/linearProblem.h:
```diff
@@ -0,0 +1,60 @@
+    std::map<std::string, std::shared_ptr<MipConstraint>> constraints_;
+
+    std::unique_ptr<MipSolution> solution_;
+    std::unique_ptr<MipObjective> objective_;
```
[ @sylvlecl @pet-mit @flomnes ]
This is an implementation detail. Please move this to the implementation.

@JasonMarechal25 `std::vector` + linear search ? @sylvlecl suggests `std::unordered_map` for performance.

sylvlecl commented on src/solver/optim/ortoolsImpl/linearProblem.cpp:
```diff
@@ -0,0 +1,141 @@
+
+Api::MipVariable* OrtoolsLinearProblem::addIntVariable(double lb, double ub, const std::string& name)
+{
+    auto* mpVar = mpSolver_->MakeIntVar(lb, ub, name);
```
Add check on duplicate name before this line.

We'd better throw an exception here.

sylvlecl commented on src/solver/optim/ortoolsImpl/linearProblem.cpp:
```diff
@@ -0,0 +1,141 @@
+    mpSolver_->EnableOutput();
+    auto status = mpSolver_->Solve(*param_);
+
+    std::map<Api::MipVariable*, double> solution;
```
Let's be careful to not use pointers as the key : they will not ensure a stable sort order from one execution to the other.

payetvin replied:
Using a map with std::string as key will need MipSolution to have another container for MipVariable ptr
Function like `getOptimalValue` take a MipVariable as arg

pet-mit replied:
I agree, we can store values in a map indexed with the mipVariable name (which is unique and accessible through MipVariable.getName())

sylvlecl reviewed (commented):
Few comments from collective review

payetvin committed:
comments

payetvin committed:
format

payetvin committed:
has bounds mother class

payetvin committed:
add getname

payetvin committed:
name for constraint

payetvin committed:
move member to ortools impl

payetvin committed:
Scenario for problem data

payetvin committed:
linearproblemfiller files

payetvin committed:
remove linear filler ortools

payetvin committed:
linear problem builder

payetvin committed:
pointer

pet-mit commented on src/solver/optim/api/include/antares/solver/optim/api/linearProblemBuilder.h:
```diff
@@ -0,0 +1,40 @@
+class LinearProblemBuilder
+{
+public:
+    virtual void init(std::vector<LinearProblemFiller*> fillers) = 0;
```
this can be the constructor

pet-mit commented on src/solver/optim/api/include/antares/solver/optim/api/linearProblemFiller.h:
```diff
@@ -0,0 +1,38 @@
+class LinearProblemFiller
+{
+public:
+    virtual void addVariable(LinearProblem* problem, LinearProblemData* data) = 0;
```
```suggestion
    virtual void addVariables(LinearProblem* problem, LinearProblemData* data) = 0;
```

pet-mit commented on src/solver/optim/api/include/antares/solver/optim/api/linearProblemFiller.h:
```diff
@@ -0,0 +1,38 @@
+public:
+    virtual void addVariable(LinearProblem* problem, LinearProblemData* data) = 0;
+    virtual void addConstraints(LinearProblem* problem, LinearProblemData* data) = 0;
+    virtual void addObjective(LinearProblem* problem, LinearProblemData* data) = 0;
```
```suggestion
    virtual void addObjectiveCoefficients(LinearProblem* problem, LinearProblemData* data) = 0;
```

pet-mit commented on src/solver/optim/api/include/antares/solver/optim/api/mipObjective.h:
```diff
@@ -0,0 +1,43 @@
+    virtual double getCoefficient(MipVariable* var) = 0;
+
+    virtual bool getMaximization() = 0;
+    virtual bool getMinimization() = 0;
```
```suggestion
    virtual bool isMaximization() = 0;
    virtual bool isMinimization() = 0;
```

pet-mit commented on src/solver/optim/api/include/antares/solver/optim/api/mipSolution.h:
```diff
@@ -0,0 +1,49 @@
+    virtual MipStatus getStatus() = 0;
+    virtual double getObjectiveValue() = 0;
+    virtual double getOptimalValue(const MipVariable& var) const = 0;
+    virtual std::vector<double> getOptimalValue(const std::vector<MipVariable>& vars) const = 0;
```
```suggestion
    virtual std::vector<double> getOptimalValues(const std::vector<MipVariable>& vars) const = 0;
```

pet-mit commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/linearProblem.h:
```diff
@@ -0,0 +1,64 @@
+    operations_research::MPSolverParameters* param_;
+
+    std::map<std::string, std::shared_ptr<Api::MipVariable>> variables_;
+    std::map<std::string, std::shared_ptr<Api::MipConstraint>> constraints_;
```
does this work?
```suggestion
    std::map<std::string, std::unique_ptr<Api::MipVariable>> variables_;
    std::map<std::string, std::unique_ptr<Api::MipConstraint>> constraints_;
```

pet-mit commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/mipConstraint.h:
```diff
@@ -0,0 +1,59 @@
+
+    operations_research::MPConstraint* mpConstraint_;
+
+    std::string name;
```
I'm not sure you need to store the name, it's already stored in the mpConstraint

pet-mit commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/mipSolution.h:
```diff
@@ -0,0 +1,57 @@
+
+    std::map<const Api::MipVariable*, double> solution_;
+
+    OrtoolsMipObjective* objective_;
```
you only need to store the optimal value of the objective

pet-mit commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/mipVariable.h:
```diff
@@ -0,0 +1,58 @@
+
+    operations_research::MPVariable* mpVar_;
+
+    std::string name;
```
no need to store this

pet-mit commented on src/solver/optim/ortoolsImpl/linearProblem.cpp:
```diff
@@ -0,0 +1,163 @@
+        logs.error() << "This variable already exists: " << name;
+        throw std::bad_function_call();
+    }
+    auto* mpVar = mpSolver_->MakeNumVar(lb, ub, name);
```
ortools has a "makeVar" function that has an extra boolean argument to say if the var is integer. you can implement a "makeVar" private function in this class like that one, and use it in both makeIntVar and makeNumVar, in order to avoid code duplication

pet-mit commented on src/solver/optim/ortoolsImpl/linearProblem.cpp:
```diff
@@ -0,0 +1,163 @@
+
+Api::MipSolution* OrtoolsLinearProblem::solve()
+{
+    mpSolver_->EnableOutput();
```
I think we should only enable solver output if log level is low enough (debug or trace)

payetvin replied:
done here for legacy: 
src/solver/utils/ortools_utils.cpp:357

we can add a boolean to `solve()`

pet-mit commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/mipVariable.h:
```diff
@@ -0,0 +1,58 @@
+
+    const std::string& getName() const override;
+
+    operations_research::MPVariable* getMpVar();
```
we can avoid making this public by nesting all Ortools implementation classes in one super-class (LinearProblem implementation)

pet-mit commented on src/solver/optim/ortoolsImpl/mipConstraint.cpp:
```diff
@@ -0,0 +1,88 @@
+
+void OrtoolsMipConstraint::setCoefficient(Api::MipVariable* var, double coefficient)
+{
+    auto* mpvar = dynamic_cast<OrtoolsMipVariable*>(var);
```
I think we can avoid dynamic cast by storing in the OrToolsMipConstraint a pointer to the OrToolsLinearProblem, and making the list of variables in OrtoolsLinearProblem accessible to OrtoolsMipConstraint.

then we'd have something like :
auto* mpvar = ortoolsLinearProblem.getMipVariable(var.getName());

(you would need to change the return value of OrtoolsLinearProblem::getMipVariable to a OrtoolsMipVariable, which is OK from a polymorphism point-of-view)

but I'm not really sure it's worth the hassle. What do you think?

guilpier-code replied:
If we do what you suggest, would not we have a performance issue ?
I mean : we would need to search for a var in a list for each call to **setCoefficient(...)**, right ? 

That said, we're in a constraint here, and the constraint could contain the list of variables associated to it.
In these conditions, the search would be much faster.

pet-mit replied:
actually we'd have to search through all the variables of the LinearProblem, since we don't know beforehand which variables are in the constraintes (we create an "empty" constraint first, then add variable coefficients one by one

pet-mit commented on src/solver/optim/ortoolsImpl/mipObjective.cpp:
```diff
@@ -0,0 +1,83 @@
+namespace Antares::Solver::Optim::OrtoolsImpl
+{
+
+OrtoolsMipObjective::OrtoolsMipObjective(operations_research::MPObjective* objective):
```
I think that having a MipObjective object is only useful if we want a LinearProblem.getObjective() function; meaning that users will have to go through that object in order to change the optimization sense or the objective coefficients. If not, I would remove the MipObjective object and put all this code in the LinearProblem implementation

payetvin committed:
use std string for map

payetvin committed:
use ptr

payetvin committed:
format

payetvin committed:
Comments 1

payetvin committed:
Comments 2

payetvin committed:
Comments 3

a-zakir commented on src/solver/optim/api/include/antares/solver/optim/api/hasBounds.h:
```diff
@@ -0,0 +1,39 @@
+namespace Antares::Solver::Optim::Api
+{
+
+class HasBounds
```
the name is misleading

a-zakir replied:
BoundsManager?

a-zakir commented on src/solver/optim/api/include/antares/solver/optim/api/linearProblem.h:
```diff
@@ -0,0 +1,53 @@
+
+    virtual void setObjectiveCoefficient(MipVariable* var, double coefficient) = 0;
+
+    virtual double getObjectiveCoefficient(MipVariable* var) = 0;
```
``
```suggestion
    virtual double getObjectiveCoefficient(const MipVariable const* var) = 0;
```

a-zakir commented on src/solver/optim/api/include/antares/solver/optim/api/linearProblem.h:
```diff
@@ -0,0 +1,53 @@
+public:
+    virtual MipVariable* addNumVariable(double lb, double ub, const std::string& name) = 0;
+    virtual MipVariable* addIntVariable(double lb, double ub, const std::string& name) = 0;
+    virtual MipVariable* getVariable(const std::string& name) = 0;
```
```suggestion
    virtual MipVariable* getVariable(const std::string& name) const = 0;
```

a-zakir commented on src/solver/optim/api/include/antares/solver/optim/api/linearProblem.h:
```diff
@@ -0,0 +1,53 @@
+    virtual MipVariable* getVariable(const std::string& name) = 0;
+
+    virtual MipConstraint* addConstraint(double lb, double ub, const std::string& name) = 0;
+    virtual MipConstraint* getConstraint(const std::string& name) = 0;
```
```suggestion
    virtual MipConstraint* getConstraint(const std::string& name) const = 0;
```

payetvin committed:
unique_ptr

a-zakir commented on src/solver/optim/api/include/antares/solver/optim/api/linearProblemData.h:
```diff
@@ -0,0 +1,47 @@
+    bool hasScalarData(const std::string& key);
+    double getScalarData(const std::string& key, unsigned scenario);
+    bool hasTimedData(const std::string& key);
+    std::vector<double>& getTimedData(const std::string& key, unsigned scenario);
```
const& return

a-zakir commented on src/solver/optim/api/include/antares/solver/optim/api/linearProblemBuilder.h:
```diff
@@ -0,0 +1,40 @@
+class LinearProblemBuilder
+{
+public:
+    virtual void LinearProblemBuilder(std::vector<LinearProblemFiller*> fillers) = 0;
```
pure virtual constructor?

payetvin committed:
don't store name for mip var and constraint

payetvin committed:
only store objective value in mip solution

payetvin committed:
addVariable() to de duplicate code

flomnes commented on src/solver/optim/ortoolsImpl/linearProblem.cpp:
```diff
@@ -41,17 +41,18 @@ OrtoolsLinearProblem::OrtoolsLinearProblem(bool isMip, const std::string& solver
     if (variables_.contains(name))
     {
         logs.error() << "This variable already exists: " << name;
         throw std::bad_function_call();
```
Don't use predefined exceptions, use your own types

flomnes commented on src/solver/optim/ortoolsImpl/linearProblem.cpp:
```diff
@@ -67,31 +68,18 @@ Api::MipVariable* OrtoolsLinearProblem::addNumVariable(double lb,
     return variables_[name].get();
```
Here you can get `try_emplace`'s returned iterator at l. 68, so you don't need to search for what you just inserted.

See https://en.cppreference.com/w/cpp/container/map/try_emplace

flomnes commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/mipSolution.h:
```diff
@@ -0,0 +1,57 @@
+private:
+    Api::MipStatus responseStatus_;
+
+    std::map<std::string, std::pair<const Api::MipVariable*, double>> solution_;
```
It might be better for performance to keep flat structures (ie `vector`)

flomnes commented on src/solver/optim/api/include/antares/solver/optim/api/mipConstraint.h:
```diff
@@ -0,0 +1,37 @@
+namespace Antares::Solver::Optim::Api
+{
+
+class MipConstraint: public virtual HasBounds, public virtual HasName
```
Do you have any reason to use `virtual` inheritance ?

payetvin replied:
Avoid diamond inheritance

JasonMarechal25 replied:
There is no diamond here though. HasName and Has bound don't share ancestry

JasonMarechal25 replied:
https://stackoverflow.com/questions/4543537/should-i-almost-always-use-virtual-inheritance#comment4987107_4543549

flomnes commented on src/solver/optim/api/include/antares/solver/optim/api/mipObjective.h:
```diff
@@ -0,0 +1,43 @@
+    virtual double getCoefficient(MipVariable* var) = 0;
+
+    virtual bool isMaximization() = 0;
+    virtual bool isMinimization() = 0;
```
```suggestion
    virtual double getCoefficient(MipVariable* var) const = 0;

    virtual bool isMaximization() const = 0;
    virtual bool isMinimization() const = 0;
```

flomnes commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/linearProblem.h:
```diff
@@ -0,0 +1,67 @@
+    std::map<std::string, std::unique_ptr<Api::MipConstraint>> constraints_;
+
+    std::unique_ptr<Api::MipSolution> solution_;
+    std::unique_ptr<Api::MipObjective> objective_;
```
Use concrete types ?

flomnes commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/mipSolution.h:
```diff
@@ -0,0 +1,57 @@
+                       const std::map<std::string, std::pair<Api::MipVariable*, double>>& solution,
+                       double objectiveValue);
+
+    ~OrtoolsMipSolution() = default;
```
```suggestion
    virtual ~OrtoolsMipSolution() = default;
```

flomnes commented on src/solver/optim/api/include/antares/solver/optim/api/hasBounds.h:
```diff
@@ -0,0 +1,39 @@
+
+    virtual void setBounds(double lb, double ub) = 0;
+
+    virtual double getLb() = 0;
+    virtual double getUb() = 0;
```
add `const` qualifier

flomnes commented on src/solver/optim/api/include/antares/solver/optim/api/mipVariable.h:
```diff
@@ -0,0 +1,34 @@
+namespace Antares::Solver::Optim::Api
+{
+
+class MipVariable: public virtual HasBounds, public virtual HasName
```
`virtual` inheritance probably not necessary

flomnes commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/linearProblem.h:
```diff
@@ -0,0 +1,67 @@
+#include <ortools/linear_solver/linear_solver.h>
+#include <string>
+
+#include <antares/solver/optim/api/linearProblem.h>
```
You may just fwd-declare `operations_research::MPSolverParameters` and `operations_research::MPSolver`. This header is kind of heavy so we avoid including it unless necessary (`ie` in cpp files)

flomnes commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/mipConstraint.h:
```diff
@@ -0,0 +1,57 @@
+
+#pragma once
+
+#include <ortools/linear_solver/linear_solver.h>
```
Forward declare `operations_research::MPConstraint` instead

flomnes commented on src/solver/optim/ortoolsImpl/linearProblem.cpp:
```diff
@@ -0,0 +1,155 @@
+        solution.try_emplace(var->name(), pair);
+    }
+
+    auto* mpObjective = dynamic_cast<OrtoolsMipObjective*>(objective_.get());
+    double objectiveValue = mpObjective->getMpObjective()->Value();
```
Interface leak. If you only need the objective value, then add a `double value()` virtual method to the interface.

flomnes commented on src/solver/optim/ortoolsImpl/mipSolution.cpp:
```diff
@@ -0,0 +1,90 @@
+  double objectiveValue):
+    objectiveValue_(objectiveValue)
+{
+    // Only store non-zero values
+    for (const auto& varAndValue: solution)
+    {
+        if (abs(varAndValue.second.second) > 1e-6) // TODO: is this tolerance OK?
+        {
+            solution_.insert(varAndValue);
+        }
+    }
```
I think it's not a good idea

payetvin replied:
Don't remove zeros, memory impact should be negligible

flomnes commented on src/solver/optim/ortoolsImpl/mipSolution.cpp:
```diff
@@ -0,0 +1,90 @@
+
+double OrtoolsMipSolution::getOptimalValue(const Api::MipVariable* var) const
+{
+    return solution_.contains(var->getName()) ? solution_.at(var->getName()).second : 0;
```
Not necessarily a good idea to assume most solution variables will be 0

payetvin replied:
what should we do ?

payetvin replied:
simplify, return solution[var]

payetvin committed:
expose AddVariable

payetvin committed:
use return of try_emplace

payetvin committed:
comments

payetvin committed:
implement objective in linear solver

payetvin committed:
remove mipObjective files

payetvin committed:
forward declaratoins

payetvin committed:
Bool verbose solver

payetvin committed:
include string hasname.h

payetvin committed:
includes

payetvin committed:
format

payetvin committed:
concrete types in linear problem

payetvin committed:
Revert "concrete types in linear problem"

This reverts commit 132118f1b11de031c3ae2b993cf95be582ddc81d.

payetvin committed:
include linear_solver.h only where needed

payetvin committed:
const functions

payetvin committed:
const functions mip var

payetvin committed:
const linear

payetvin committed:
format

payetvin committed:
custom exception when elem already exists

payetvin commented on src/solver/optim/ortoolsImpl/mipSolution.cpp:
```diff
@@ -0,0 +1,90 @@
+    // Only store non-zero values
+    for (const auto& varAndValue: solution)
+    {
+        if (abs(varAndValue.second.second) > 1e-6) // TODO: is this tolerance OK?
```
std::abs

payetvin committed:
Simplify solution, store zeros

payetvin committed:
Move enum conversion to linear problem

payetvin committed:
format

payetvin committed:
use concrete types

payetvin committed:
base for tus

payetvin committed:
basic add var

payetvin committed:
try fix windows compile

JasonMarechal25 commented on src/solver/optim/api/include/antares/solver/optim/api/hasBounds.h:
```diff
@@ -0,0 +1,39 @@
+namespace Antares::Solver::Optim::Api
+{
+
+class HasBounds
```
By convention we name interfaces (pure virtual class without member) ISomething. I suggest we follow this convention

a-zakir committed:
try fix

payetvin committed:
correct error for constraitn

payetvin committed:
Merge branch 'feature/1.1-api-optim' of https://github.com/AntaresSimulatorTeam/Antares_Simulator into feature/1.1-api-optim

payetvin committed:
comment some tests

payetvin committed:
fix tests

payetvin committed:
get var

payetvin committed:
objective

payetvin committed:
Merge branch 'develop' into feature/1.1-api-optim

payetvin committed:
constraint coeff tu

payetvin committed:
Merge remote-tracking branch 'origin/develop' into feature/1.1-api-optim

payetvin committed:
virtual desctuctors

payetvin committed:
add createproblem func and try solve in tu

payetvin committed:
remove ptr

payetvin committed:
default fixture destructor

payetvin committed:
test infeasible problem

payetvin committed:
final destructor

payetvin committed:
try solving simple problem

payetvin committed:
test optimal values

payetvin committed:
code smells

payetvin committed:
excpetion

payetvin committed:
format

payetvin committed:
remove destrcutor

payetvin committed:
more code smells

payetvin committed:
mip variable bounds test

payetvin committed:
mip constraint bounds test

payetvin committed:
format

payetvin committed:
check throw

payetvin committed:
other check

payetvin committed:
check throw coeff

payetvin committed:
test exception

payetvin committed:
Catch exception

payetvin committed:
remove virtual for inheritance

payetvin committed:
Use i before class name to signify interface

payetvin committed:
test null ptr

payetvin committed:
format

payetvin committed:
test vector null values

JasonMarechal25 commented on src/solver/optim/ortoolsImpl/linearProblem.cpp:
```diff
@@ -0,0 +1,215 @@
+        return Api::MipStatus::FEASIBLE;
+    case operations_research::MPSolver::ResultStatus::UNBOUNDED:
+        return Api::MipStatus::UNBOUNDED;
+    default:
```
It's often a good idea to throw or at least log something in default cases. When a new enum value is added this function will return error but we won't have any information why.

JasonMarechal25 commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/mipConstraint.h:
```diff
@@ -0,0 +1,60 @@
+
+    const std::string& getName() const override;
+
+    ~OrtoolsMipConstraint() final = default;
```
Is final intended and meaningful ?

payetvin replied:
Sonarcloud detected it as a code smell, I think it makes sense as no class should inherit from this one

JasonMarechal25 replied:
In this case you can set the class final, not just the destructor.

JasonMarechal25 commented on src/solver/optim/api/include/antares/solver/optim/api/linearProblemData.h:
```diff
@@ -0,0 +1,47 @@
+
+class LinearProblemData
+{
+public:
```
It seems some functions could be marked as const

payetvin committed:
Start adding doxygen

pet-mit commented on src/solver/optim/api/include/antares/solver/optim/api/linearProblem.h:
```diff
@@ -0,0 +1,73 @@
+    /// Create a variable and specify if it's an integer
+    virtual IMipVariable* addVariable(double lb, double ub, bool integer, const std::string& name)
+      = 0;
+    /// Wrapper for addVariable with integer = false
```
```suggestion
    /// Create a continuous variable
```

pet-mit commented on src/solver/optim/api/include/antares/solver/optim/api/linearProblem.h:
```diff
@@ -0,0 +1,73 @@
+      = 0;
+    /// Wrapper for addVariable with integer = false
+    virtual IMipVariable* addNumVariable(double lb, double ub, const std::string& name) = 0;
+    /// Wrapper for addVariable with integer = true
```
```suggestion
    /// Create an integer variable
```

pet-mit commented on src/solver/optim/api/include/antares/solver/optim/api/linearProblem.h:
```diff
@@ -0,0 +1,73 @@
+    virtual IMipConstraint* addConstraint(double lb, double ub, const std::string& name) = 0;
+    virtual IMipConstraint* getConstraint(const std::string& name) = 0;
+
+    /// Set the objective for a given variable
```
```suggestion
    /// Set the objective coefficient for a given variable
```

pet-mit commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/mipSolution.h:
```diff
@@ -0,0 +1,56 @@
+private:
+    Api::MipStatus responseStatus_;
+
+    std::map<std::string, std::pair<const Api::IMipVariable*, double>> solution_;
```
why do we need to store the IMIPVariable pointers?
I think the following might be enough, because variables are uniquely identified by their names:
```suggestion
    std::map<std::string, double> solution_;
```

pet-mit commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/mipVariable.h:
```diff
@@ -0,0 +1,59 @@
+
+    ~OrtoolsMipVariable() final = default;
+
+    explicit OrtoolsMipVariable(operations_research::MPVariable*);
```
I think we should definitely remove this by doing what i suggested (add a pointer to the OrtoolsLinearProblem inside every OrtoolsMipConstraint instance)

however we'll have to watch out later for eventual performance issues causes by the extra variable lookup

pet-mit commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/linearProblem.h:
```diff
@@ -0,0 +1,79 @@
+    bool isMinimization() const override;
+    bool isMaximization() const override;
+
+    Api::IMipSolution* solve(bool verboseSolver) override;
```
This would offer us more felxibility to use polymorphism:
```suggestion
    OrtoolsMipVariable* addNumVariable(double lb, double ub, const std::string& name) override;
    OrtoolsMipVariable* addIntVariable(double lb, double ub, const std::string& name) override;
    OrtoolsMipVariable* addVariable(double lb,
                                   double ub,
                                   bool integer,
                                   const std::string& name) override;
    OrtoolsMipVariable* getVariable(const std::string& name) override;

    OrtoolsMipConstraint* addConstraint(double lb, double ub, const std::string& name) override;
    OrtoolsMipConstraint* getConstraint(const std::string& name) override;

    void setObjectiveCoefficient(Api::IMipVariable* var, double coefficient) override;
    double getObjectiveCoefficient(const Api::IMipVariable* var) const override;

    void setMinimization() override;
    void setMaximization() override;

    bool isMinimization() const override;
    bool isMaximization() const override;

    OrtoolsMipSolution* solve(bool verboseSolver) override;
```

payetvin committed:
Remove nullptr from test objective values

payetvin committed:
Add warning for default error status

payetvin committed:
Doc

payetvin committed:
Simplify mipSolution vector

payetvin committed:
Use ortools types

payetvin committed:
Add a simple problem to demo

payetvin committed:
format

payetvin committed:
doxygen

payetvin committed:
final class

payetvin committed:
const methods

guilpier-code commented on src/solver/CMakeLists.txt:
```diff
@@ -19,6 +19,7 @@ add_subdirectory(simulation)
 add_subdirectory(ts-generator)
 add_subdirectory(utils)
 add_subdirectory(variable)
+add_subdirectory(optim)
```
_**This is a draft remark / question**_ (you don't have to answer it) :

Both directories **solver/optimisation** and **solver/optim** wear more or less the same name.
Will **solver/optimisation** disappear at some point ?

pet-mit replied:
the idea was to avoid interaction between the old and the new codes, so we created this new namespace. yes, I think "optimisation" will disappear at some point

guilpier-code replied:
Changing **solver/optim**  into **solver/modeler** would be quick and make things more clear.

payetvin replied:
I agree

guilpier-code replied:
Done

payetvin committed:
Merge branch 'develop' into feature/1.1-api-optim

guilpier-code commented on src/solver/optim/api/include/antares/solver/optim/api/linearProblemData.h:
```diff
@@ -0,0 +1,47 @@
+class LinearProblemData
+{
+public:
+    unsigned getTimeResolutionInMinutes();
```
Function **getTimeResolutionInMinutes** : 
Time resolution can be misleading and can refer to the duration of the problem resolution.
**getTimeGranularityInMinutes** ?

guilpier-code commented on src/solver/optim/api/include/antares/solver/optim/api/linearProblemData.h:
```diff
@@ -0,0 +1,47 @@
+class LinearProblemData
+{
+public:
+    unsigned getTimeResolutionInMinutes();
```
Question on **getTimeResolutionInMinutes** : 
Why **in minutes** ?
The smallest time granularity in Antares Simulator is hour, so I can't see a usage for a time amount expressed in minutes

pet-mit replied:
I think we better not limit it to the hour right now ; one of the main objectives of andromede being to give the user the biggest amount of flexibility possible

guilpier-code replied:
As principles **YAGNI** and **KISS** say, we should not make assumptions for what will come in the future.
Our concern should be to make the current code the more clear and simple as possible.
Changes will come in time, if and when we need them.

pet-mit replied:
limiting to the hour would also be making an assumption, because the API is supposed to allow future use cases that are not possible today in antares. and it would be more painful to change back, to allow minutes. the other way around (the current version) allows setting the granularity to 1 minute or 1 hour, has no overcost, and is more future-proof.

guilpier-code replied:
The shortest time interval in Antares is the **hour**, that's not an assumption, but a fact.
Now, what will happen if we leave this **getTimeResolutionInMinute** as it is, and know that there is no such **minutes** in Antares Simulator ? Well it's just confusing, and it will certainly make someone waste time.

Besides, changing back the name of a function at some point and because we need it is not time consuming.

pet-mit replied:
I checked with the POs and they did not seem to have an opinion about this. so if you prefer, we can go back to 1-hour resolution in the meantime

guilpier-code commented on src/tests/src/solver/optim/api/testApiOrtoolsLinearProblem.cpp:
```diff
@@ -0,0 +1,242 @@
+
+#define WIN32_LEAN_AND_MEAN
+
+#include <ortools/linear_solver/linear_solver.h>
```
Apparently we need `#include <ortools/linear_solver/linear_solver.h>` here in the unit tests.
Why is that ?
Indeed, ortools linear solver interface (I mean the dependency interface) should be needed inside the ortools implementation of our problem interface (class **OrtoolsLinearProblem**), but not elsewhere.

guilpier-code commented on src/solver/optim/api/include/antares/solver/optim/api/linearProblem.h:
```diff
@@ -0,0 +1,74 @@
+ * along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+ */
+
+#pragma once
```
We should not hesitate to refer to ortools in the header/source file names : **ortoolsLinearProblem .h/.cpp**

guilpier-code commented on src/tests/src/solver/optim/api/testApiOrtoolsLinearProblem.cpp:
```diff
@@ -0,0 +1,242 @@
+    void createProblemMaximize();
+};
+
+void Fixture::createProblemMaximize()
```
We expect **a**, **b**, and **c** to be of same type, as they naturally follow one another in a kind of logical sequence.
But actually, they are different things (have a different type).
These objects should wear more expressive names, for example : 
```C++
auto* var1 = pb->addNumVariable(0, 1, "var1");
auto* var2 = pb->addNumVariable(0, 1, "var2");
auto* someConstraint = pb->addConstraint(1, 1, "someConstraint");
```

guilpier-code commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/linearProblem.h:
```diff
@@ -0,0 +1,79 @@
+class OrtoolsLinearProblem final: public Api::ILinearProblem
+{
+public:
+    OrtoolsLinearProblem(bool isMip, const std::string& solverName);
```
About : 
```C++
OrtoolsLinearProblem(bool isMip, const std::string& solverName);
```
We have a boolean **isMip** as an argument of the constructor here.
It's used to build either a MIP or a LP problem internally.
On the other hand, we have a list of MIP variables and a list of MIP constraints as data members.
Question :
- If we have a LP problem, isn't that strange and leads to troubles having this **LP** problem working with **MIP** variables or **MIP** constraints ?
- For now, we're about to use class **OrtoolsLinearProblem** only with `isMip == false`. Do we not make things more complex than it should be at this stage ?
- Even if we're sure MILP problem will come into play at some point, should we not make 2 classes instead of one (one for LP, one for MILP), and possibly have the common code either in a common parent class, or in a polymorphic data member (see pattern strategy) ? The goal would be to have smaller classes, reducing one their responsibilities perimeter, as required by software quality.

flomnes replied:
> If we have a LP problem, isn't that strange and leads to troubles having this LP problem working with MIP variables or MIP constraints ?

UB or better, error. This is related to how MPSolver handles MIP/LP problems. We might make the decision later in the execution flow (is there any MIP variable ? If so, use MIP, otherwise use LP), at the price of a bit more complexity.

> For now, we're about to use class OrtoolsLinearProblem only with isMip == false. Do we not make things more complex than it should be at this stage ?

No, we'll 100% want to use MIP problems later

> Even if we're sure MILP problem will come into play at some point, should we not make 2 classes instead of one (one for LP, one for MILP), and possibly have the common code either in a common parent class, or in a polymorphic data member (see pattern strategy) ? The goal would be to have smaller classes, reducing one their responsibilities perimeter, as required by software quality.

Duplicated classes wouldn't be easy to use/maintain in my opinion.

guilpier-code replied:
> Duplicated classes wouldn't be easy to use/maintain in my opinion.

This is not what I suggest. On the contrary, I suggest putting the common code in a separate code unit to avoid duplication. But we'll see that in time.

guilpier-code commented on src/tests/src/solver/optim/api/testApiOrtoolsLinearProblem.cpp:
```diff
@@ -0,0 +1,242 @@
+
+    pb->setObjectiveCoefficient(a, 1);
+    pb->setObjectiveCoefficient(b, 1);
+    pb->setMaximization();
```
`pb->setMaximization();` : 
Do we need to "maximize" for a problem that small ?

flomnes replied:
I'm not sure I understand, can you elaborate ?

guilpier-code replied:
> I'm not sure I understand, can you elaborate ?

I mean : whatever means "maximizing" a problem, do we need to maximize a problem made for a several unit tests, especially if that problem is very small ?
Said in another way, if we remove the maximization, do we loose something ?

flomnes replied:
Maximize is meant to tell the solver something along the lines of

> Find x such that b.x is **maximal** under the given set of constraints & bounds

By calling `setMaximization`, we just set the sign (+/-) of the objective function `x -> b.x`. It is not intended to expand the problem, whatever that means.

guilpier-code commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/linearProblem.h:
```diff
@@ -0,0 +1,79 @@
+ * along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+ */
+
+#pragma once
```
We should not hesitate to refer to ortools in the header/source file names : **ortoolsLinearProblem .h/.cpp**.
(the parent class is contained in a pair of files that wears the same name, which can be confusing)

guilpier-code commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/linearProblem.h:
```diff
@@ -0,0 +1,79 @@
+
+    OrtoolsMipVariable* addNumVariable(double lb, double ub, const std::string& name) override;
+    OrtoolsMipVariable* addIntVariable(double lb, double ub, const std::string& name) override;
+    OrtoolsMipVariable* addVariable(double lb,
```
Should the **addVariable(...)** method with 4 arguments not be private, as it's called from **addNumVariable** and **addIntVariable**, and therefore does not seem to be required from outside ?
It would simplify this class's interface.

pet-mit replied:
this, in fact, was the original idea behind this method. but then we thought with @payetvin that there was no reason to hide it since it allows the user a little more flexibility (you can see that google made the choice to keep it public in MPSolver too). but if you think that it is confusing, we can hide it.

guilpier-code replied:
Then we should make this method **private**, and if it turns out we need it as a public method, we  change.
This will be easy to make this quick change.
In the meantime, we keep the class interface the most simple we can.

guilpier-code replied:
As a suggestion, please see [this PR](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2385)

guilpier-code replied:
Done and merge from #2385

guilpier-code commented on src/solver/optim/ortoolsImpl/linearProblem.cpp:
```diff
@@ -0,0 +1,215 @@
+    params_.SetIntegerParam(MPSolverParameters::PRESOLVE, 0);
+}
+
+class ElemAlreadyExists: public std::exception
```
**ElemAlreadyExists** : Do we need this class ?
Indeed, the **what()** function carries a message, but another more specific and clearer message is already logged just before throwing the exception.
So we could either : 
- log the error message and throw a simple **std::exception**
- only throwing the  **std::exception** with the message inside : 

```c++
throw std::exception("This variable already exists: " + name);
```

flomnes replied:
Throwing a generic exception is bad, it doesn't allow the caller to determine what happened. For example,
```cpp
void doStuff()
{
    // ...
    if (notInitialized)
    {
        throw std::exception("Stuff is not initialized");
    }
    // ...
    if (solutionNotFound)
    {
        throw std::exception("Solution was not found");
    }
}

void caller()
{
    try
    {
        doStuff()
    }
    catch (const std::exception& ex)
    {
        // Can't determine programmatically what went wrong in doStuff
    }
}
```
Deriving classes from e.g `std::runtime_error` is the preferred alternative.
```cpp
#include <stdexcept>
#include <string>

class NotInitializedException : public std::runtime_error {
public:
    NotInitializedException()
        : std::runtime_error("Stuff is not initialized") {}
};

class SolutionNotFoundException : public std::runtime_error {
public:
    SolutionNotFoundException()
        : std::runtime_error("Solution was not found") {}
};

void doStuff()
{
    // ...
    if (notInitialized)
    {
        throw NotInitializedException();
    }
    // ...
    if (solutionNotFound)
    {
        throw SolutionNotFoundException();
    }
}

void caller()
{
    try
    {
        doStuff();
    }
    catch (const NotInitializedException& ex)
    {
        // Handle the case where stuff was not initialized
        std::cerr << "Initialization error: " << ex.what() << std::endl;
    }
    catch (const SolutionNotFoundException& ex)
    {
        // Handle the case where the solution was not found
        std::cerr << "Solution error: " << ex.what() << std::endl;
    }
    catch (const std::exception& ex)
    {
        // Handle other general exceptions
        std::cerr << "General error: " << ex.what() << std::endl;
    }
}
```

guilpier-code replied:
ok, I see your point.
Then, should we log the error (`logs.error()`) and also throw an exception (as it is done now), knowing that the exception will be caught and its message printed ?
I guess we don't need to log **BEFORE** throwing the exception.

flomnes replied:
It depends if we expect the caller to catch the exception and log it. If so, then we don't need to send anything to the logs, effectively decoupling ourselves from `antares::logs` completely.

guilpier-code commented on src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/linearProblem.h:
```diff
@@ -0,0 +1,79 @@
+    operations_research::MPObjective* objective_;
+    operations_research::MPSolverParameters params_;
+
+    std::map<std::string, std::unique_ptr<OrtoolsMipVariable>> variables_;
+    std::map<std::string, std::unique_ptr<OrtoolsMipConstraint>> constraints_;
```
About : 
```c++
std::map<std::string, std::unique_ptr<OrtoolsMipVariable>> variables_;
std::map<std::string, std::unique_ptr<OrtoolsMipConstraint>> constraints_;
```
These 2 data members are private and use class **OrtoolsMipVariable** or **OrtoolsMipConstraint**.

Using **OrtoolsMipVariable** or  **OrtoolsMipConstraint** objects internally (or privately if you like), makes the internal code more complex than it could be by using **ortools** native types.

If we use **ortools** native types, we don't need these 2 data members  (**variables_** & **constraints_**) : these lists are already stored in **mpSolver_**.

For example, in **OrtoolsLinearProblem::addVariable(...)**, we could have : 
```C++
OrtoolsMipVariable* OrtoolsLinearProblem::addVariable(double lb,
                                                      double ub,
                                                      bool integer,
                                                      const std::string& name)
{
    if (mpSolver_.LookupVariableOrNull(name))
    {
        logs.error() << "This variable already exists: " << name;
        throw ElemAlreadyExists();
    }

    auto* mpVar = mpSolver_->MakeVar(lb, ub, integer, name);

    if (!mpVar)
    {
        logs.error() << "Couldn't add variable to Ortools MPSolver: " << name;
    }

    return std::make_unique<OrtoolsMipVariable>(mpVar).get();
}
```
This function looks simpler like this.

Some other simplification suggestions in the same spirit are done in associated **.cpp** file

flomnes replied:
```cpp
    return std::make_unique<OrtoolsMipVariable>(mpVar).get();
```
This function would return a dangling pointer 👎🏻

guilpier-code replied:
ok, but if you agree with the spirit of the remark, we can find another solution.

guilpier-code replied:
ok.
Are we forced to return a pointer to **OrtoolsMipVariable** here ?
Could we not return a **OrtoolsMipVariable** ? In that case, it would be simple : 
```c++
OrtoolsMipVariable OrtoolsLinearProblem::addVariable(...)
{
    ....
    return mpVar;
}
```

Otherwise , we still could do : 
```
...
auto var_to_return = std::make_unique<OrtoolsMipVariable>(mpVar);
return var_to_return.get();
```
or even find another solution.

flomnes replied:
Covariance doesn't apply here since e.g `OrtoolsMipVariable` isn't related to `operations_research::MPVariable`.

flomnes replied:
```cpp
auto var_to_return = std::make_unique<OrtoolsMipVariable>(mpVar);
return var_to_return.get();
``` 
is completely equivalent, since `var_to_return` is destructed at the end of scope.

guilpier-code replied:
> Covariance doesn't apply here since e.g `OrtoolsMipVariable` isn't related to `operations_research::MPVariable`.

What if **OrtoolsMipVariable** has constructor taking a **MPVariable** as an argument ?

guilpier-code replied:
> ```c++
> auto var_to_return = std::make_unique<OrtoolsMipVariable>(mpVar);
> return var_to_return.get();
> ```
> 
> is completely equivalent, since `var_to_return` is destructed at the end of scope.

Then we have the same problem in the **solve(...)** function, of which code is currently :
```C++
solution_ = std::make_unique<OrtoolsMipSolution>(solution, status, objective_->Value());
return solution_.get();
```

flomnes replied:
> > ```c++
> > auto var_to_return = std::make_unique<OrtoolsMipVariable>(mpVar);
> > return var_to_return.get();
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > is completely equivalent, since `var_to_return` is destructed at the end of scope.
> 
> Then we have the same problem in the **solve(...)** function, of which code is currently :
> 
> ```c++
> solution_ = std::make_unique<OrtoolsMipSolution>(solution, status, objective_->Value());
> return solution_.get();
> ```

No, since `solution_` is a (private) data member of class `OrtoolsLinearProblem`. As long as the instance of `OrtoolsLinearProblem`, pointer `solution_.get()` is guaranteed to remain valid (unless it is `reset`/`released`).

guilpier-code replied:
ok, sorry I missed it.

guilpier-code commented on src/solver/optim/ortoolsImpl/linearProblem.cpp:
```diff
@@ -0,0 +1,215 @@
+    return Api::MipStatus::MIP_ERROR;
+}
+
+OrtoolsMipSolution* OrtoolsLinearProblem::solve(bool verboseSolver)
```
In the same spirit as for [this comment](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2286#discussion_r1741760630) (in the associated **.h** header), this **solve(...)** method could be simpler using **ortools** native types as most as possible : 
```c++
OrtoolsMipSolution* OrtoolsLinearProblem::solve(bool verboseSolver)
{
    if (verboseSolver)
    {
        mpSolver_->EnableOutput();
    }

    auto mpStatus = mpSolver_->Solve(params_);
    return std::make_unique<OrtoolsMipSolution>(mpSolver_).get();
}
```
This function is now simple. 
Indeed, some task are now delegated to class **OrtoolsMipSolution** : it's more the mission of class **OrtoolsMipSolution** to deal with **solution** details, not the job of **OrtoolsLinearProblem**

flomnes replied:
```cpp
    return std::make_unique<OrtoolsMipSolution>(mpSolver_).get();
```
dangling pointer 👎🏻

flomnes replied:
```cpp
    return std::make_unique<OrtoolsMipSolution>(mpSolver_).get();
```
dangling pointer 👎🏻

guilpier-code replied:
So I guess a solution would be to return a smart pointer to **OrtoolsMipSolution** (and have the parent abstract function returning a smart pointer on the parent abstract class), don't you think ?

flomnes replied:
We could return a `std::shared_ptr` but it doesn't help much.

guilpier-code replied:
> We could return a `std::shared_ptr` but it doesn't help much.

Where is the problem with smart pointers ?

flomnes replied:
Performance & cycles

guilpier-code replied:
**Summary of this conversation** : 
We are currently forced to keep 2 lists of (**std::unique_ptr**) pointers : 
- **variables_** : pointer to **OrtoolsMipVariable**
- **constraints_** : pointers to **OrtoolsMipConstraint**

These lists are required for member functions `addVariable(...)`, `getVariable(...,)` to return a **raw** pointer to **OrtoolsMipVariable** with respect to covariance related to base class **ILinearProblem** (same for constraints).
These lists are "copies" of **MPVariable** objects (resp. **MPConstraint**). 

A subject could be to simplify code by removing these lists if we think they're not required.
To do that, instead of a raw pointer, these functions could return a freshly created **std** smart pointer.
A problem could that using smart pointers that way it may lower performances, so we should test it.

We should create a ticket for this : make the change and test it.

flomnes replied:
See #2367

guilpier-code replied:
See also [this ticket](https://github.com/AntaresSimulatorTeam/Antares_Simulator/issues/2386)

guilpier-code commented on src/solver/optim/ortoolsImpl/linearProblem.cpp:
```diff
@@ -0,0 +1,215 @@
+    return objective_->maximization();
+}
+
+static Api::MipStatus convertStatus(operations_research::MPSolver::ResultStatus& status)
```
About `convertStatus(...)` function : 
if we agree on [this comment](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2286/files?show-viewed-files=true&file-filters%5B%5D=#r1741827781), this function should be defined and used in source file containing class **OrtoolsMipSolution**.

guilpier-code replied:
Done in PR #2374 and merged here

guilpier-code commented on src/tests/src/solver/optim/api/testApiOrtoolsLinearProblem.cpp:
```diff
@@ -0,0 +1,242 @@
+    auto pb = std::make_unique<OrtoolsImpl::OrtoolsLinearProblem>(true, "sirius");
+}
+
+BOOST_FIXTURE_TEST_CASE(basicLinearProblemAdd, Fixture)
```
Unit tests **basicLinearProblemAdd** : 
There is no test here.
We should add a test.

If we do so, we should separate tests on adding a variable and adding a constraint : unit tests should be as small as possible and test one thing only.
Unit tests good practice [there](https://brightsec.com/blog/unit-testing-best-practices/), part 1 and 3.

guilpier-code replied:
Done in #2377 and merged here

guilpier-code commented on src/tests/src/solver/optim/api/testApiOrtoolsLinearProblem.cpp:
```diff
@@ -0,0 +1,242 @@
+    auto pb = std::make_unique<OrtoolsImpl::OrtoolsLinearProblem>(true, "sirius");
+}
+
+BOOST_FIXTURE_TEST_CASE(basicLinearProblemAdd, Fixture)
```
**Renaming** : 
basicLinearProblemAdd ==> **add_a_variable_to_a_problem___check_variable_exists**
The test title is a small & clear sentence

Unit test good practice for names [here](https://medium.com/@stefanovskyi/unit-test-naming-conventions-dd9208eadbea)

guilpier-code replied:
Done in #2377 and merged here

guilpier-code commented on src/tests/src/solver/optim/api/testApiOrtoolsLinearProblem.cpp:
```diff
@@ -0,0 +1,242 @@
+    pb->addConstraint(0, 1, "c");
+}
+
+BOOST_FIXTURE_TEST_CASE(linearProblemGetAndConstraintSetCoeff, Fixture)
```
unit test : **linearProblemGetAndConstraintSetCoeff**

The test makes too many things, we should split it.
As a result, the name of the test is unclear : indeed, it tries to express too many things in a few words.

If you try to guess what the test does without looking at its code, well... it's very hard.

guilpier-code replied:
Done in #2377 and merged here

guilpier-code commented on src/tests/src/solver/optim/api/testApiOrtoolsLinearProblem.cpp:
```diff
@@ -0,0 +1,242 @@
+    BOOST_CHECK_THROW(cons->getCoefficient(nullptr), std::bad_cast);
+}
+
+bool correctMessage(const std::exception& ex)
```
Renaming : **correctMessage(...)** ==> **expectedMessage(...)**

guilpier-code replied:
Done in #2377 and merged here

guilpier-code commented on src/tests/src/solver/optim/api/testApiOrtoolsLinearProblem.cpp:
```diff
@@ -0,0 +1,242 @@
+
+BOOST_FIXTURE_TEST_CASE(objectiveCoeff, Fixture)
+{
+    auto* var = pb->addVariable(0, 1, true, "a");
```
`auto* var = pb->addVariable(0, 1, true, "a");` : 
As said earlier, this method with 4 arguments should be **private**.
As a consequence, it's impossible to test it.
It doesn't matter, it's not meant to be.
Class / objects should be tested through their interface. 
Testing classes thoroughly by unit testing each method is not a good practice : unit testing is about testing behavior, not methods.
About this, please see [this article](https://levelup.gitconnected.com/dont-unit-test-methods-test-behaviours-bf42c8498b65)

guilpier-code commented on src/tests/src/solver/optim/api/testApiOrtoolsLinearProblem.cpp:
```diff
@@ -0,0 +1,242 @@
+
+BOOST_FIXTURE_TEST_CASE(infeasibleProblem, Fixture)
+{
+    auto* a = pb->addIntVariable(0, 10, "a");
```
If we want to test only how the problem reacts to resolution when it's infeasible, then we could use a even more simple problem : one variable (always positive) and one constraint (always negative), not even sure that we need a cost for variables.
This will make the test even shorter.

guilpier-code replied:
Done in #2377 and merged here

guilpier-code commented on src/tests/src/solver/optim/api/testApiOrtoolsLinearProblem.cpp:
```diff
@@ -0,0 +1,242 @@
+    BOOST_CHECK_EQUAL(solution->getOptimalValue(nullptr), 0);
+}
+
+BOOST_AUTO_TEST_CASE(simpleProblem)
```
unit test : **simpleProblem**

There is no test here.
This makes this "test" useless, as as it won't ever raise a testing error.

guilpier-code replied:
Removed in #2377 and merged here

guilpier-code commented on src/solver/optim/api/include/antares/solver/optim/api/hasName.h:
```diff
@@ -0,0 +1,37 @@
+{
+public:
+    virtual ~IHasName() = default;
+    virtual const std::string& getName() const = 0;
```
About **getName()** : 
We could rename it into **name()**, shorter and expressive enough.

pet-mit replied:
about that, I think we should have a common convention for getters / setters and stick to it in all the code

guilpier-code commented on src/solver/optim/ortoolsImpl/mipSolution.cpp:
```diff
@@ -0,0 +1,84 @@
+
+    try
+    {
+        return solution_.at(var->getName());
```
`solution_` : 
same question again : do we need this data members ?
Can we not extract the information we need directly from **ortools MP** data ?

guilpier-code replied:
Same as [this](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2286#discussion_r1751517355)

guilpier-code replied:
See [this ticket](https://github.com/AntaresSimulatorTeam/Antares_Simulator/issues/2386)

guilpier-code commented on src/solver/optim/ortoolsImpl/linearProblem.cpp:
```diff
@@ -0,0 +1,215 @@
+                                                      bool integer,
+                                                      const std::string& name)
+{
+    if (variables_.contains(name))
```
if it turns out we cant avoid hosting **variables_** and **constraints_** : 
**Question** : 
Do the following instructions : 
- ` if (constraints_.contains(name)) `
- `if (!mapIteratorPair.second)`

check the same thing ?

If it turns out they do, a more simplify form of this function : 
```C++
OrtoolsMipVariable* OrtoolsLinearProblem::addVariable(double lb,
                                                      double ub,
                                                      bool integer,
                                                      const std::string& name)
{
    if (variables_.contains(name))
    {
        logs.error() << "This variable already exists: " << name;
        throw ElemAlreadyExists();
    }

    auto* mpVar = mpSolver_->MakeVar(lb, ub, integer, name);

    if (!mpVar)
    {
        logs.error() << "Couldn't add variable to Ortools MPSolver: " << name;
    }

    const auto& pair = variables_.try_emplace(name, std::make_unique<OrtoolsMipVariable>(mpVar));
    return pair.first->second.get(); // <<name, var>, bool>
}
```

guilpier-code replied:
Done in #2374

guilpier-code committed:
API modeller : apply remarks (#2374)

This PR applies several remarks from #2286.
Mostly simplifications :
- code has shrunk a bit (less lines).
- class **OrtoolsLinearProblem** is smaller now, because : 
- **solve** method was simplified a bit, and some tasks were delegated
to class **OrtoolsMipSolution**
  - **add** (variables & constraints) methods were a bit simplified

guilpier-code committed:
Modeler API : apply remarks of #2286 on unit tests (#2377)

The goal of this PR is to improve existing unit tests on modeler's API.
To do that we : 
- shortened unit tests. It mostly implied splitting existing unit tests
into smaller one
- removed unit test with no check
- tried to simplify fixtures and make them more clear
- make LP problem used in these tests more simple when possible.

guilpier-code committed:
Modeler api : rename folder "optim" into "modeler"  (#2378)

This PR is about modeler API.
This PR is a new step into taking remarks of PR #2286 into account. 
2 things are done here : 
- very small simplification in getting the objective value after problem
resolution
- but mostly we rename the folder **optim** into **modeler**, more
adapted.

payetvin committed:
Fix conflicts

guilpier-code committed:
Modeler API : make addVariable with 4 args private (#2385)

a-zakir committed:
Fix/sonar issues from 2286 (#2384)

pet-mit committed:
Merge branch 'develop' into feature/1.1-api-optim
